internal/xds/rbac/rbac_engine.go

```diff
@@ -15,68 +15,218 @@
+	case v3rbacpb.RBAC_ALLOW:
+		action = allow
+	case v3rbacpb.RBAC_DENY:
+		action = deny
+	case v3rbacpb.RBAC_LOG:
+		return nil, errors.New("unsupported action")
+	}
+
+	policies := make(map[string]*policyMatcher, len(policy.Policies))
 	for name, config := range policy.Policies {
 		matcher, err := newPolicyMatcher(config)
 		if err != nil {
 			return nil, err
 		}
 		policies[name] = matcher
 	}
-	return &Engine{policies: policies}, nil
+	return &engine{
+		policies: policies,
+		action:   action,
+	}, nil
 }
 
-// RPCData wraps data pulled from an incoming RPC that the RBAC engine needs to
-// find a matching policy.
-type RPCData struct {
-	// MD is the HTTP Headers that are present in the incoming RPC.
-	MD metadata.MD
-	// PeerInfo is information about the downstream peer.
-	PeerInfo *peer.Peer
-	// FullMethod is the method name being called on the upstream service.
-	FullMethod string
-	// DestinationPort is the port that the RPC is being sent to on the
-	// server.
-	DestinationPort uint32
-	// DestinationAddr is the address that the RPC is being sent to.
-	DestinationAddr net.Addr
-	// PrincipalName is the name of the downstream principal. If set, the URI
-	// SAN or DNS SAN in that order is used from the certificate, otherwise the
-	// subject field is used. If unset, it applies to any user that is
-	// authenticated.
-	PrincipalName string
-}
+var errPolicyNotFound = errors.New("a matching policy was not found")
 
-// FindMatchingPolicy determines if an incoming RPC matches a policy. On a
-// successful match, it returns the name of the matching policy and a true
-// boolean to specify that there was a matching policy found.
-func (r *Engine) FindMatchingPolicy(data *RPCData) (string, bool) {
+// findMatchingPolicy determines if an incoming RPC matches a policy. On a
+// successful match, it returns the name of the matching policy and a nil error
+// to specify that there was a matching policy found.  It returns an error in
+// the case of not finding a matching policy.
+func (r *engine) findMatchingPolicy(rpcData *rpcData) (string, error) {
 	for policy, matcher := range r.policies {
-		if matcher.match(data) {
-			return policy, true
+		if matcher.match(rpcData) {
+			return policy, nil
+		}
+	}
+	return "", errPolicyNotFound
+}
+
+// newRPCData takes a incoming context (should be a context representing state
+// needed for server RPC Call with metadata, peer info (used for source ip/port
+// and TLS information) and connection (used for destination ip/port) piped into
+// it) and the method name of the Service being called server side and populates
+// an rpcData struct ready to be passed to the RBAC Engine to find a matching
+// policy.
+func newRPCData(ctx context.Context) (*rpcData, error) {
+	// The caller should populate all of these fields (i.e. for empty headers,
+	// pipe an empty md into context).
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, errors.New("missing metadata in incoming context")
```

Comment:
This will work. However, are you suggesting that we change it in case what is piped into the context in grpc.Server changes at a later time? Personally, I think since metadata and peer info are so crucial to everything on grpc.Server, that it will never be not be in the context passed into the interceptor. If this is a big deal for you/you have a strong opinion about this, let me know and I can change it back.